sdk/canton/community/participant/src/main/scala/com/digitalasset/canton/participant/admin/PackageUpgradeValidator.scala

```diff
@@ -8,35 +8,80 @@ import com.daml.daml_lf_dev.DamlLf.Archive
 import com.daml.error.DamlError
 import com.daml.lf.archive.Decode
 import com.daml.lf.data.Ref
-import com.daml.lf.language.Ast
+import com.daml.lf.language.Util.dependenciesInTopologicalOrder
+import com.daml.lf.language.{Ast, LanguageVersion}
 import com.daml.lf.validation.{TypecheckUpgrades, UpgradeError}
 import com.daml.logging.entries.LoggingValue.OfString
 import com.digitalasset.canton.ledger.error.PackageServiceErrors.{InternalError, Validation}
 import com.digitalasset.canton.logging.LoggingContextWithTrace.implicitExtractTraceContext
 import com.digitalasset.canton.logging.{LoggingContextWithTrace, NamedLoggerFactory, NamedLogging}
+import com.digitalasset.canton.participant.admin.PackageUpgradeValidator.PackageMap
 import com.digitalasset.canton.participant.admin.PackageUploader.ErrorValidations
 import com.digitalasset.canton.tracing.TraceContext
-import scalaz.std.either.*
-import scalaz.std.option.*
+import com.digitalasset.canton.util.EitherTUtil
+import scalaz.std.either._
+import scalaz.std.option._
 import scalaz.std.scalaFuture.futureInstance
-import scalaz.syntax.traverse.*
+import scalaz.syntax.traverse._
 
 import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits.infixOrderingOps
+
+object PackageUpgradeValidator {
+  type PackageMap = Map[Ref.PackageId, (Ref.PackageName, Ref.PackageVersion)]
+}
 
 class PackageUpgradeValidator(
-    getPackageMap: TraceContext => Map[Ref.PackageId, (Ref.PackageName, Ref.PackageVersion)],
+    getPackageMap: TraceContext => PackageMap,
     getLfArchive: TraceContext => Ref.PackageId => Future[Option[Archive]],
     val loggerFactory: NamedLoggerFactory,
 )(implicit executionContext: ExecutionContext)
     extends NamedLogging {
 
   def validateUpgrade(
+      upgradingPackages: List[(Ref.PackageId, Ast.Package)]
+  )(implicit
+      loggingContext: LoggingContextWithTrace
+  ): EitherT[Future, DamlError, Unit] = {
+    val upgradingPackagesMap = upgradingPackages.toMap
+    val deps = dependenciesInTopologicalOrder(upgradingPackages.map(_._1), upgradingPackagesMap)
```

Comment:
isn't `deps` misleading? According to the docs of that method,
`Returns the `pkgIds` and all its dependencies in topological order.`
so `packagesInTopologicalOrder` would be better suited no?